src/instructlab/eval/ragas.py

```diff
@@ -0,0 +1,212 @@
+        self.student_model = student_model
+        self.run_config = run_config
+        self.openai_client = openai_client
+
+    def run(
+        self,
+        dataset: List[Sample] | Path,
+        student_model: ModelConfig | None = None,
+        run_config: RunConfig | None = None,
+        openai_client: OpenAIClient | None = None,
+    ) -> EvaluationResult:
+        """
+        Evaluates the quality of model responses against a graded rubric.
+
+        When the `dataset` lacks the `response` field, then `student_model` must be provided
+        in order to generate the answers.
+
+        Args:
+            dataset (List[Sample] | Path):
+                Can be either a list of `Sample` objects or a path to a jsonl file containing
+                records matching `Sample`.
+            student_model: (StudentModelConfig):
+                When this parameter is provided, we'll attempt to use the described model in order to
+                generate the responses from the given list of questions.
+            run_config (RunConfig | None, optional):
+                Configuration to use when running evaluations. If none is provided, then
+                a default one is created containing extremely permissive settings when handling
+                timeouts. This is because by default, OpenAI tier-1 usage accounts have very high
+                rate limits resulting in heavy throttling during evaluations.
+            openai_client (openai.Client | None, optional):
+                The client to use when generating questions from the student model, must be compatible with the OpenAI API.
+                This field is required when `student_model` is provided.
+
+        Returns:
+            EvaluationResult: The results of all evaluations performed by Ragas
+        """
+        student_model = student_model if student_model else self.student_model
+        run_config = run_config if run_config else self.run_config
+        openai_client = openai_client if openai_client else self.openai_client
+
+        # ensure we are in the dataframe format
+        input_df = None
+        if isinstance(dataset, list):
+            input_df = DataFrame(dataset)
+        elif isinstance(dataset, Path):
+            input_df = read_json(dataset, orient="records", lines=True)
+        else:
+            raise TypeError(f"invalid type of dataset: {type(dataset)}")
+
+        # this should never happen, but pylint is not smart enough to detect it
+        assert input_df is not None
+
+        need_to_generate_questions = "response" not in input_df.columns
+        if need_to_generate_questions and (not student_model or not openai_client):
+            raise ValueError(
+                "provided dataset doesn't contain the model `response`, but either `student_model` or `openai_client` wasn't provided for inference"
+            )
+
+        # if the student model was provided then we always generate regardless
+        if student_model:
```

Comment:
Can we combine this conditional and the one above it. It seems if `needs_to_generate_questions` is set then we'd run `_generate_answers_from_model()`.

The reason I say that is I accidentally passed in input with responses, but also the student model and the openai_client when I was using this PR. What would you want to happen in that case? Ignore the responses column and ask the student model the questions and get the responses?

Reply:
Yeah the idea is that when the student model is provided then it will automatically be used to generate responses. But the reason why we can't combine these two conditionals is that there's a separate state where the student model was not passed in, but there isn't sufficient information to perform the evaluation (e.g. the `response` field is missing), then we error out because we cannot continue.

Reply:
I do think that this can be designed better, but let's focus on getting this PR merged. Then we can focus on improvements afterwards.

Reply:
Agree we can always tweak later